src/Microsoft.Extensions.Primitives/StringSegment.cs

```diff
@@ -30,22 +54,61 @@ public StringSegment(string buffer, int offset, int length)
             Length = length;
         }
 
+        /// <summary>
+        /// The <see cref="string"/> that acts as a buffer for this segment.
+        /// </summary>
         public string Buffer { get; }
 
+        /// <summary>
+        /// The offset within the Buffer at wich the segment starts.
+        /// </summary>
         public int Offset { get; }
 
+        /// <summary>
+        /// The length of the segment starting from the offset defined by the segment.
+        /// </summary>
         public int Length { get; }
 
+        /// <summary>
+        /// The <see cref="string"/> represented by this segment.
+        /// </summary>
         public string Value
         {
-            get { return HasValue ? Buffer.Substring(Offset, Length) : null; }
+            get
+            {
+                if (!HasValue)
+                {
+                    return null;
+                }
+                else
+                {
+                    // Typically, the original string was wrapped in a StringSegment in order to make it fit
+                    // some method or member signature, so we just unwrap it.
+                    if (Offset == 0 && Buffer.Length == Length)
+                    {
+                        return Buffer;
+                    }
+                    else
+                    {
+                        return Buffer.Substring(Offset, Length);
```

Comment:
I believe substring already does this optimization. Measure before you go writing extra code.